src/svcutilcore/src/Microsoft/Tools/ServiceModel/SvcUtil/Options.cs

```diff
@@ -404,12 +403,67 @@ private void AddMscorlib(Dictionary<string, Type> foundCollectionTypes)
                 }
             }
 
-            private void AddServiceModelLib(Dictionary<string, Type> foundCollectionTypes)
+            private void LoadSMReferenceAssembly()
             {
-                Assembly serviceModelLib = typeof(ChannelFactory).Assembly;
-                if (!_parent._referencedAssemblies.Contains(serviceModelLib))
+                IList<string> referencedAssembliesArgs = _arguments.GetArguments(Options.Cmd.SMReference);
+                if (referencedAssembliesArgs != null && referencedAssembliesArgs.Count > 0)
                 {
-                    AddReferencedTypesFromAssembly(serviceModelLib, foundCollectionTypes);
+                    string smassembly = referencedAssembliesArgs[0];
+                    if(smassembly.LastIndexOf("System.ServiceModel.Primitives") == -1)
+                    {
+                        ToolConsole.WriteError("Need pass the right path of System.ServiceModel.Primitives for smreference parameter");
+                        throw new ArgumentException("Invalid smreference value");
+                    }
+
+                    string smpassembly = smassembly.Replace("System.ServiceModel.Primitives", "System.Private.ServiceModel");
+                    //for some lowercase path
+                    smpassembly = smpassembly.Replace("system.servicemodel.primitives", "system.private.servicemodel");
+
+                    int refplace = smassembly.LastIndexOf("ref");
+                    if(refplace > 0 )
+                    {
+                        smassembly = smassembly.Remove(refplace, 3).Insert(refplace, "lib");
+                        refplace = smpassembly.LastIndexOf("ref");
+                        smpassembly = smpassembly.Remove(refplace, 3).Insert(refplace, "lib");
+                        int libplace = smpassembly.LastIndexOf("lib");
+                        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                        {
```

Comment:
Use System.IO.Path.DirectorySeparatorChar and don't do platform detection.

Reply:
Actually, ignore that, I just realized there is unix and win in the paths.

---
In reply to: [224268609](https://github.com/dotnet/wcf/pull/3188#discussion_r224268609) [](ancestors = 224268609)